src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/Streaming/DataStreamWriterTests.cs

```diff
@@ -59,5 +63,205 @@ public void TestSignaturesV2_3_X()
+            TestForeachWriter foreachWriter,
+            bool foreachThrows,
+            int partitions,
+            long expectedCSVFiles,
+            long expectedExceptionFiles,
+            IEnumerable<int> expectedOutput)
+        {
+            // Temporary folder the TestForeachWriter will write to.
+            using var dstTempDirectory = new TemporaryDirectory();
+
+            foreachWriter.WritePath = dstTempDirectory.Path;
+
+            // Read streamInputPath, repartitions data into `partitions`, then
+            // calls TestForeachWriter on the data.
+            DataStreamWriter dsw = _spark
+                .ReadStream()
+                .Schema("id INT")
+                .Json(streamInputPath)
+                .Repartition(partitions)
+                .WriteStream()
+                .Foreach(foreachWriter);
+
+            // Trigger the stream batch once.
+            if (foreachThrows)
+            {
+                Assert.Throws<Exception>(
+                    () => dsw.Trigger(Trigger.Once()).Start().AwaitTermination());
+            }
+            else
+            {
+                dsw.Trigger(Trigger.Once()).Start().AwaitTermination();
+            }
+
+            // Verify that TestForeachWriter created a unique .csv when
+            // ForeachWriter.Open was called on each partitionId.
+            Assert.Equal(
+                expectedCSVFiles,
+                Directory.GetFiles(dstTempDirectory.Path, "*.csv").Length);
+
+            // Only if ForeachWriter.Process(Row) throws an exception, will
+            // ForeachWriter.Close(Exception) create a file with the
+            // .exeception extension.
+            Assert.Equal(
+                expectedExceptionFiles,
+                Directory.GetFiles(dstTempDirectory.Path, "*.exception").Length);
+
+            // Read in the *.csv file(s) generated by the TestForeachWriter.
+            // If there are multiple input files, sorting by "id" will make
+            // validation simpler. Contents of the *.csv will only be populated
+            // on successful calls to the ForeachWriter.Process method.
+            DataFrame foreachWriterOutputDF = _spark
+                .Read()
+                .Schema("id INT")
+                .Csv(dstTempDirectory.Path)
+                .Sort("id");
+
+            // Validated expected *.csv data.
+            Assert.Equal(
+                    expectedOutput.Select(i => new object[] { i }),
+                    foreachWriterOutputDF.Collect().Select(r => r.Values));
+        }
+
+        [Serializable]
+        private class TestForeachWriter : IForeachWriter
+        {
+            // Mark the StreamWriter as ThreadStatic. If there are multiple Tasks
+            // running this ForeachWriter, on the same Worker, and in parallel, then
+            // it may be possible that they may use the same StreamWriter object. This
+            // may cause an unintended side effect of a Task writing the output to a
```

Comment:
Can we just remove this side effect? You can use locks such that there is always only one `StreamWriter` per file.

Reply:
I also see this comment ` A single instance of this class is responsible of all the data generated by a single task`. Do we need to handle thread issue here?

Reply:
Originally I this a static field because the `StreamWriter` was not marked as `[Serializable]`.  But since we are creating the `StreamWriter` object in the `open` call, marking the field as `[NonSerialized]` addresses the issue.

Reply:
> I also see this comment ` A single instance of this class is responsible of all the data generated by a single task`. Do we need to handle thread issue here?

If all the fields of the ForeachWriter are non static then I don't think there will be issues, but if there are static fields then I think it'll be up to the user to determine how to handle them.